internal/discoverer/nacos.go

```diff
@@ -3,95 +3,83 @@ package discoverer
 import (
 	"errors"
 	"fmt"
+	"hash"
+	"hash/crc32"
+	"net/url"
 	"strconv"
 	"strings"
 	"sync"
 
-	"github.com/nacos-group/nacos-sdk-go/model"
-
-	"github.com/api7/apisix-seed/internal/utils"
-
-	"github.com/nacos-group/nacos-sdk-go/clients"
-	"github.com/nacos-group/nacos-sdk-go/vo"
-
 	"github.com/api7/apisix-seed/internal/conf"
 	"github.com/api7/apisix-seed/internal/core/comm"
+	"github.com/api7/apisix-seed/internal/utils"
+	"github.com/nacos-group/nacos-sdk-go/clients"
 	"github.com/nacos-group/nacos-sdk-go/clients/naming_client"
 	"github.com/nacos-group/nacos-sdk-go/common/constant"
+	"github.com/nacos-group/nacos-sdk-go/model"
+	"github.com/nacos-group/nacos-sdk-go/vo"
 )
 
 func init() {
 	Discoveries["nacos"] = NewNacosDiscoverer
 }
 
+func serviceID(service string, args map[string]string) string {
+	serviceId := fmt.Sprintf("%s@%s@%s", args["namespace_id"], args["group_name"], service)
+	return serviceId
+}
+
 type NacosDiscoverer struct {
-	weight int
-	// nacos client default config
-	ClientConfig constant.ClientConfig
-	// nacos server configs
-	ServerConfigs []constant.ServerConfig
-	// nacos naming clients
-	namingClients map[string]naming_client.INamingClient
+	timeout uint64
+	weight  int
+	// nacos server configs, grouping by authentication information
+	ServerConfigs map[string][]constant.ServerConfig
+	// nacos naming clients, grouping by authentication information
+	namingClients map[string][]naming_client.INamingClient
 
 	params map[string]*vo.SubscribeParam
-	cache  map[string]Service
+	cache  map[string]*Service
 	mu     sync.Mutex
 
+	crc hash.Hash32
+
 	watchCh chan *comm.Watch
 }
 
 func NewNacosDiscoverer(disConfig interface{}) (Discoverer, error) {
 	config := disConfig.(*conf.Nacos)
-	timeout := config.Timeout
 
-	clientConfig := constant.ClientConfig{
-		// compatible with past timeout configurations
-		TimeoutMs:           uint64(timeout.Connect + timeout.Read + timeout.Send),
-		NamespaceId:         config.Namespace,
-		Username:            config.Username,
-		Password:            config.Password,
-		NotLoadCacheAtStart: true,
-	}
-
-	serverConfigs := make([]constant.ServerConfig, len(config.Host))
-	for i, host := range config.Host {
-		schemaEnd, ipEnd := strings.Index(host, "://"), strings.LastIndex(host, ":")
+	serverConfigs := make(map[string][]constant.ServerConfig)
+	for _, host := range config.Host {
+		u, err := url.Parse(host)
+		if err != nil {
+			return nil, err
+		}
 
 		port := 80 // default port
```

Comment:
The value of the default port should depend on the scheme

Reply:
I found in the Nacos documentation and code that the default port is 8848 , and is not related to scheme. I think it may need to be changed to 8848 here .

Reply:
OK